src/dp_nat.c

```diff
@@ -296,6 +297,184 @@ void dp_nat_chg_ip(struct dp_flow *df, struct rte_ipv4_hdr *ipv4_hdr,
+		else if (icmp_hdr->icmp_type == DP_ICMPV6_ECHO_REPLY)
+			icmp_hdr->icmp_type = RTE_IP_ICMP_ECHO_REPLY;
+		else
+			return 0;
+		dp_calculate_icmp_checksum(icmp_hdr, rte_be_to_cpu_16(ipv4_hdr->total_length) - ((ipv4_hdr->version_ihl & 0x0F) * 4));
+	break;
+	default:
+		return DP_ERROR;
+	}
+
+	return DP_OK;
+}
+
+int dp_nat_chg_ipv4_to_ipv6_hdr(struct dp_flow *df, struct rte_mbuf *m, uint8_t *ipv6_addr)
+{
+	struct rte_ether_hdr *eth_hdr;
+	struct rte_ipv4_hdr *ipv4_hdr;
+	struct rte_ipv6_hdr *ipv6_hdr;
+	struct rte_icmp_hdr *icmp_hdr;
+	struct rte_udp_hdr *udp_hdr;
+	struct rte_tcp_hdr *tcp_hdr;
+	uint32_t src_ipv4;
+	uint8_t l4_proto;
+
+	eth_hdr = rte_pktmbuf_mtod(m, struct rte_ether_hdr *);
+	eth_hdr->ether_type = rte_cpu_to_be_16(RTE_ETHER_TYPE_IPV6);
+
+	ipv4_hdr = dp_get_ipv4_hdr(m);
+	src_ipv4 = ipv4_hdr->src_addr;
+	l4_proto = ipv4_hdr->next_proto_id;
+
+	// Adjust the packet data to fit IPv6
+	if (rte_pktmbuf_prepend(m, sizeof(struct rte_ipv6_hdr) - sizeof(struct rte_ipv4_hdr)) == NULL)
+		return DP_ERROR;
+
+	// Move the Ethernet header to just before the IPv6 header
+	memmove(rte_pktmbuf_mtod(m, uint8_t *), eth_hdr, sizeof(struct rte_ether_hdr));
+
+	// Setup the new IPv6 header
+	ipv6_hdr = dp_get_ipv6_hdr(m);
+	ipv6_hdr->vtc_flow = rte_cpu_to_be_32((6 << 28) | (ipv4_hdr->type_of_service << 20));
+	ipv6_hdr->payload_len = rte_cpu_to_be_16(rte_pktmbuf_data_len(m) - (uint16_t)sizeof(struct rte_ether_hdr) - (uint16_t)sizeof(struct rte_ipv6_hdr));
+	ipv6_hdr->proto = l4_proto;
+	ipv6_hdr->hop_limits = ipv4_hdr->time_to_live;
+
+	// Set the source and destination IPv6 addresses
+	rte_memcpy(ipv6_hdr->src_addr, DP_NAT64_PREFIX, 12);
+	*(uint32_t *)&ipv6_hdr->src_addr[12] = src_ipv4;
+	rte_memcpy(ipv6_hdr->dst_addr, ipv6_addr, DP_IPV6_ADDR_SIZE);
+	rte_memcpy(df->dst.dst_addr6, ipv6_addr, DP_IPV6_ADDR_SIZE);
+
+	m->packet_type &= ~RTE_PTYPE_L3_MASK;
+	m->ol_flags |= RTE_MBUF_F_TX_IPV6;
+	m->l2_len = sizeof(struct rte_ether_hdr);
+	m->l3_len = sizeof(struct rte_ipv6_hdr);
+	m->l4_len = 0;
+
+	switch (df->l4_type) {
+	case IPPROTO_TCP:
+		tcp_hdr = (struct rte_tcp_hdr *)(ipv6_hdr + 1);
+		tcp_hdr->cksum = 0;
+		m->ol_flags |= RTE_MBUF_F_TX_TCP_CKSUM;
+		m->l4_len = DP_TCP_HDR_LEN(tcp_hdr);
+		break;
+	case IPPROTO_UDP:
+		udp_hdr = (struct rte_udp_hdr *)(ipv6_hdr + 1);
+		udp_hdr->dgram_cksum = 0;
+		m->ol_flags |= RTE_MBUF_F_TX_UDP_CKSUM;
+		m->l4_len = sizeof(struct rte_udp_hdr);
+		break;
+	case IPPROTO_ICMP:
+		df->l4_type = IPPROTO_ICMPV6;
+		m->l4_len = sizeof(struct rte_icmp_hdr);
+		ipv6_hdr->proto = IPPROTO_ICMPV6;
+		icmp_hdr = (struct rte_icmp_hdr *)(ipv6_hdr + 1);
+		icmp_hdr->icmp_code = 0;
+		icmp_hdr->icmp_cksum = 0;
+
+		if (icmp_hdr->icmp_type == RTE_IP_ICMP_ECHO_REQUEST)
+			icmp_hdr->icmp_type = DP_ICMPV6_ECHO_REQUEST;
+		if (icmp_hdr->icmp_type == RTE_IP_ICMP_ECHO_REPLY)
+			icmp_hdr->icmp_type = DP_ICMPV6_ECHO_REPLY;
+		icmp_hdr->icmp_cksum = rte_ipv6_udptcp_cksum(ipv6_hdr, icmp_hdr);
```

Comment:
This piece of code is similar to the other direction, yet it does not use `else if ` and `else error`. Not sure if intentional

Reply:
Was not intentional. The blocks in both directions should look and behave similar. Corrected.